### PR TITLE
Add finetuning.py script for automated retraining on (old/new) blended data

### PR DESCRIPTION
**Pull request by aayushmnit**
image

The finetune.py script outlines the following process for continued update of the model:

1. Downloading old/new training data and preprocessing it (placeholders)
2. Blending old/new training data (implemented in `data_blender`)
3. Downloading a seed model checkpoint (placeholder)
4. Finetuning (all layers) the seed model on blended with FastAI framework conveniences

### CODE REVIEW

File: ModelTraining/finetuning.py [hunk 1]
```diff
@@ -0,0 +1,283 @@
...
+
+def data_blender(dataPath):
+    """
+    Function to blend the original data with new data
+    
+    """
+    pos_samples  = len((dataPath/'positive').ls())
+    neg_samples  = len((dataPath/'negative').ls())
+    new_neg_samples  = len((dataPath/'new_samples').ls())
+    total_neg_samples = neg_samples + new_neg_samples
```
Comment (Molkree) [inline ModelTraining/finetuning.py:154]:
```suggestion
    neg_samples  = len((dataPath/'negative').ls())
    new_neg_samples  = len((dataPath/'new_samples').ls())
```
Currently `pos_samples` and `total_neg_samples` are not used.

File: ModelTraining/finetuning.py [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+    
+    neg_img_list = pd.Series((dataPath/'negative').ls() + (dataPath/'new_samples').ls())
```
Comment (Molkree) [inline ModelTraining/finetuning.py:156]: Can we just use `neg_samples` and `new_neg_samples` from above or do we need to read again?
```suggestion
    neg_img_list = pd.Series(neg_samples + new_neg_samples)
```

File: ModelTraining/finetuning.py [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+    
+    ## Randomly selecting neg_samples for overall list
+    new_neg_img_list = neg_img_list.sample(n=neg_samples, replace=False).values
+    
+    ## Get image name
+    new_neg_img_list = [str(item).split('/')[-1] for item in new_neg_img_list]
+    
+    ## copying all data from new samples to negative
+    for item in (dataPath/'new_samples').ls():
```
Comment (Molkree) [inline ModelTraining/finetuning.py:165]:
```suggestion
    for item in new_neg_samples:
```

File: ModelTraining/finetuning.py [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+        shutil.move(src= str(item), dst = dataPath/'negative', copy_function = shutil.copy)
+        
+    ## removing new samples directory
+    shutil.rmtree(dataPath/'new_samples')
+        
+    for item in (dataPath/'negative').ls():
```
Comment (Molkree) [inline ModelTraining/finetuning.py:171]:
```suggestion
    for item in neg_samples:
```

File: ModelTraining/finetuning.py [hunk 5]
```diff
@@ -0,0 +1,283 @@
...
+            n_mfcc=20,
+        ),
+    )
+    config.duration = 4000  # 4 sec padding or snip
+    config.resample_to = 20000  # Every sample at 20000 frequency
+    config.downmix = True
+
+    # Creating Data Loader
+    audios = (
+        AudioList.from_folder(data_folder, config=config)
```
Comment (Molkree) [inline ModelTraining/finetuning.py:235]: `dataPath` is the parameter, `data_folder` is just its default value.
```suggestion
        AudioList.from_folder(dataPath, config=config)
```

**Review by Molkree — DISMISSED** [review]

File: ModelTraining/finetuning.py [hunk 6]
```diff
@@ -0,0 +1,283 @@
...
+    '''
+    Function to load a wav file
+    '''
+    return AudioSegment.from_wav(wav_file)
+
+
+def export_wave_file(audio, begin, end, dest):
+    '''
+    Function to extract a smaller wav file based start and end duration information
+    '''
```
[the remark this answers is not in the document]
Comment (prakruti) [reply ModelTraining/finetuning.py:71]: @Molkree yep, that was a typo. Thanks for pointing it out.

File: ModelTraining/finetuning.py [hunk 7]
```diff
@@ -0,0 +1,283 @@
...
+        - abc2.wav
+        ...
+    |-new_data/
+        -somefiles0.wav
+        -somefiles1.wav
+        ...
+"""
+
+# Function to pre-process the new audio file
+def get_wave_file(wav_file):
```
[the remark this answers is not in the document]
Comment (akashmjn) [reply ModelTraining/finetuning.py:56]: Keeping this quick and dirty with placeholders for now. Can be refactored in a follow-up

File: ModelTraining/finetuning.py [hunk 8]
```diff
@@ -0,0 +1,283 @@
...
+        -modelName(.pkl file)
+        -newModelName(.pkl file)
+"""
+
+'''
+Still need to write the function
+test_model()
+if test is looking good --
+    - upload_model(data_folder/'models/'+'newModelName')
+'''
```
[the remark this answers is not in the document]
Comment (akashmjn) [reply ModelTraining/finetuning.py]: resolving keeping the placeholders functions. dont mind having the overall structure here for future reference, until individual components are filled in

File: ModelTraining/finetuning.py [hunk 9]
```diff
@@ -0,0 +1,283 @@
...
+    """
+    Function to convert new audio file containing False Negative into model-ready stream
+    Input -
+    dataPath: path to data folder
+    Output -
+    Will automatically put processed files in the 'filePath/new_samples/' folder
+    """
+
+    ## Create o/p folder 
+    local_dir = dataPath/"new_samples"
```
[the remark this answers is not in the document]
Comment (akashmjn) [reply ModelTraining/finetuning.py]: resolving as it's a pathlib.Path object

**Review by akashmjn — APPROVED** [review]
Significantly cleaned up and now good to merge
